File: 300.500/461.HammingDistance.cpp
#include <iostream>
#include <bitset>
using namespace std;

class HammingDistance{
public:
    int hammingDistance(int x, int y) {
        int z = x ^ y;
        int ret = 0;
        while(z > 0){
            z = z & (z-1);
            ret++;
        }
        return ret;
    }
public:
    int hammingDistance1(int x, int y) {
        int z = x ^ y;
        int ret = 0;
        while(z > 0){
            ret += z & 1;
            z = z >> 1;
        }
        return ret;
    }
public:
    int hammingDistance2(int x, int y) {
        bitset<32> bs(x ^ y);
        return bs.count();
    }
};

int main(int argc, char const *argv[]){
    HammingDistance cls;
    cout << cls.hammingDistance2(1,4) << " -> 2" << endl;
    cout << cls.hammingDistance2(3,1) << " -> 1" << endl;
    return 0;
}

File: go/300.500/461.HammingDistance.go
package main

import (
	"fmt"
	"math/bits"
)

func hammingDistance(x int, y int) int {
	return bits.OnesCount(uint(x ^ y))
}

func hammingDistance1(x int, y int) int {
	z := x ^ y
	ret := 0
	for z > 0 {
		ret += z & 1
		z = z >> 1
	}
	return ret
}

func hammingDistance2(x int, y int) int {
	z := x ^ y
	ret := 0
	for z > 0 {
		ret++
		z = z & (z - 1)
	}
	return ret
}

func main() {
	fmt.Println(hammingDistance(1, 4), 2)
	fmt.Println(hammingDistance(3, 1), 1)

	fmt.Println(hammingDistance1(1, 4), 2)
	fmt.Println(hammingDistance1(3, 1), 1)

	fmt.Println(hammingDistance2(1, 4), 2)
	fmt.Println(hammingDistance2(3, 1), 1)
}
